### PR TITLE
Initial PR for RKNPU execution provider

### PR DESCRIPTION
**Pull request by airockchip**
* Init

    * Support Ops:
        Conv, Relu, Clip, LeakyRelu,
        MaxPool, AveragePool, GlobalAveragePool,
        Concat, Softmax, BatchNormalization, Gemm,
        Add, Mul, Sub,
        Reshape, Squeeze, Unsqueeze,
        Flatten, Transpose,
        QLinearConv, DequantizeLinear

    * Add rknpu unittest

    * Update BUILD.md and Add RKNPU-ExecutionProvider.md

**Description**:  It is an initial PR for RKNPU execution provider. 

**Motivation and Context**

- For support Rockchip NPU hardware, such as RK1808
- RKNPU general support  fp16 and uint8/int8 quantization, but the computing power is concentrated in uint8. currently onnx only support little quantization op, so the acceleration effect is not very good

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -61,3 +61,6 @@
 [submodule "cmake/external/onnx-tensorrt"]
 	path = cmake/external/onnx-tensorrt
 	url = https://github.com/stevenlix/onnx-tensorrt.git
+[submodule "cmake/external/rknpu_ddk"]
```
Comment (jywu-msft) [inline .gitmodules:64]: I took a look at https://github.com/airockchip/rknpu_ddk. It seems like the main purpose is for distributing the rknpu library?
it's preferred to not add this as a submodule to onnxruntime repo. We don't users to pull binaries when cloning onnxruntime repo.
instead, can you list these libraries as a dependency in the BUILD instructions and direct users to download the libraries themselves?

File: BUILD.md [hunk 2]
```diff
@@ -449,6 +450,33 @@ export LD_LIBRARY_PATH=~/ComputeLibrary/build/
...
+* See [Build ARM](#ARM) below for information on building for ARM devices
+* Use gcc-linaro-6.3.1-2017.05-x86_64_aarch64-linux-gnu instead of gcc-linaro-6.3.1-2017.05-x86_64_arm-linux-gnueabihf, and modify CMAKE_CXX_COMPILER & CMAKE_C_COMPILER in tool.cmake:
+  ```
+  set(CMAKE_CXX_COMPILER aarch64-linux-gnu-g++)
+  set(CMAKE_C_COMPILER aarch64-linux-gnu-gcc)
+  ```
+
+#### Build Instructions
+##### Linux
+1. Build ONNX Runtime library and test:
```
Comment (jywu-msft) [inline BUILD.md:468]: please add instructions for fetching rknpu library.
add a build option for the location of the installed libraries and headers.

File: cmake/onnxruntime_providers.cmake [hunk 3]
```diff
@@ -480,6 +484,38 @@ if (onnxruntime_USE_NNAPI)
...
 
+if (onnxruntime_USE_RKNPU)
+  # set(CMAKE_VERBOSE_MAKEFILE ON)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-variable -Wno-unused-parameter")
+  add_definitions(-DUSE_RKNPU=1)
+  option(DNN_READ_ONNX "" ON)
+  set(DNN_CUSTOM_PROTOC_EXECUTABLE ${ONNX_CUSTOM_PROTOC_EXECUTABLE})
+  option(DNN_CMAKE_INSTALL "" OFF)
+  option(DNN_BUILD_BIN "" OFF)
+  set(RKNPU_DDK_INCLUDE_DIR ${REPO_ROOT}/cmake/external/rknpu_ddk/include)
```
Comment (jywu-msft) [inline cmake/onnxruntime_providers.cmake:495]: as mentioned previously, better to use a build option for rknpu library location.
instead of fetching via git submodule.

**Comment by jywu-msft** [comment]
Thanks for the contribution!
General comment, where possible, can you put RKNPU EP code in its own nested namespace under onnxruntime?

File: onnxruntime/core/providers/rknpu/onnx_converter.cc [hunk 4]
```diff
@@ -0,0 +1,1893 @@
...
+      supported_node_vec.push_back(i);
+    } else {
+      const auto& op = model_proto.graph().node(i).op_type();
+      LOGS_DEFAULT(INFO) << op << ": " << error_msg;
+      if (!supported_node_vec.empty()) {
+        // if (supported_node_vec.size() >= 2)
+        supported_node_vecs.push_back(supported_node_vec);
+        supported_node_vec.clear();
+      }
+      // if (op == "NonMaxSuppression" || op == "TopK")
```
Comment (jywu-msft) [inline onnxruntime/core/providers/rknpu/onnx_converter.cc:578]: is this temporary?

File: .gitmodules (hunk 1 above)
Comment (airockchip) [reply .gitmodules]: ok, I have moved rknpu_ddk out of cmake/external.

File: BUILD.md (hunk 2 above)
Comment (airockchip) [reply BUILD.md]: already update

File: cmake/onnxruntime_providers.cmake (hunk 3 above)
Comment (airockchip) [reply cmake/onnxruntime_providers.cmake]: updated

File: onnxruntime/core/providers/rknpu/onnx_converter.cc (hunk 4 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/onnx_converter.cc]: it's temporary code, I have removed it.

**Comment by airockchip** [comment]
> Thanks for the contribution!
> General comment, where possible, can you put RKNPU EP code in its own nested namespace under onnxruntime?
rknpu namespace has been added.

File: onnxruntime/core/providers/rknpu/node_attr_helper.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ public:
+  NodeAttrHelper(ONNX_NAMESPACE::NodeProto proto);
+
+  float get(const std::string& key, float def_val);
+  int get(const std::string& key, int def_val);
+  std::vector<float> get(const std::string& key, std::vector<float> def_val);
+  std::vector<int> get(const std::string& key, std::vector<int> def_val);
+  std::string get(const std::string& key, std::string def_val);
+
+  bool has_attr(const std::string& key);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/node_attr_helper.h:27]: >); [](start = 38, length = 2)

const?

File: onnxruntime/core/providers/rknpu/node_attr_helper.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ */
+class NodeAttrHelper {
+ public:
+  NodeAttrHelper(ONNX_NAMESPACE::NodeProto proto);
+
+  float get(const std::string& key, float def_val);
+  int get(const std::string& key, int def_val);
+  std::vector<float> get(const std::string& key, std::vector<float> def_val);
+  std::vector<int> get(const std::string& key, std::vector<int> def_val);
+  std::string get(const std::string& key, std::string def_val);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/node_attr_helper.h:25]: Should all of them be const methods?

File: onnxruntime/core/providers/rknpu/node_attr_helper.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace onnxruntime {
+namespace rknpu {
+
+/**
+ * Wrapping onnx::NodeProto for retrieving attribute values
+ */
+class NodeAttrHelper {
+ public:
+  NodeAttrHelper(ONNX_NAMESPACE::NodeProto proto);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/node_attr_helper.h:19]: >Nod [](start = 2, length = 3)

explicit

File: onnxruntime/core/providers/rknpu/node_attr_helper.cc [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+using std::vector;
+
+namespace onnxruntime {
+namespace rknpu {
+
+NodeAttrHelper::NodeAttrHelper(ONNX_NAMESPACE::NodeProto proto) : node_(proto) {
+}
+
+float NodeAttrHelper::get(const std::string& key, float def_val) {
+  for (int i = 0; i < node_.attribute_size(); i++) {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/node_attr_helper.cc:21]: >(int i = 0; i < no [](start = 6, length = 18)

would using for(:) loop be more readable?

File: onnxruntime/core/providers/rknpu/node_attr_helper.cc [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+//
+
+#include "node_attr_helper.h"
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+using std::string;
+using std::vector;
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/node_attr_helper.cc:12]: We always qualify 3rd party names std::string, std;;vector

File: onnxruntime/core/providers/rknpu/onnx_converter.h [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+                                const std::string& output);
+  void AddLayerQuantizeLinear(const std::string& input,
+                              const std::string& output_scale,
+                              const std::string& output_zp,
+                              const std::string& output);
+
+  void Clear();
+
+ public:
+  OnnxConverter() {}
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/onnx_converter.h:226]: >OnnxConverter() {} [](start = 2, length = 18)

What is the copy/assignment policy for the class?

File: onnxruntime/core/providers/rknpu/onnx_converter.h [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+  virtual ~OnnxConverter() { Clear(); }
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/onnx_converter.h:227]: Why introducing a virtual table when a) not overriding anthing, b) the class is not designed for inheritance

File: onnxruntime/core/providers/rknpu/onnx_converter.h [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+  // for GetSupportedNodes
+  std::map<std::string, std::vector<uint32_t>> tensor_dims_;
+
+  std::vector<void*> free_list_;  // remember free
+
+  std::pair<std::pair<int, ONNX_NAMESPACE::NodeProto>, FuseCode>
+  FindActivation(const ONNX_NAMESPACE::ModelProto& model_proto,
+                 const std::string& output_name);
+
+  std::shared_ptr<rk::nn::Tensor>
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/onnx_converter.h:48]: >std::shared_ptr<rk::nn::Tensor> [](start = 2, length = 31)

What is the purpose of the shared_ptr?

File: onnxruntime/core/providers/rknpu/onnx_converter.h [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+#include <unordered_map>
+#include <string>
+
+#include "shaper.h"
+
+#include "rknpu/rknpu_pub.h"
+
+namespace onnxruntime {
+namespace rknpu {
+class OnnxConverter {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/onnx_converter.h:20]: >lass OnnxConverter { [](start = 1, length = 20)

The class and its methods lack documentation.

File: onnxruntime/core/providers/rknpu/shaper.h [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+  void Squeeze(const std::string& input_name, const std::vector<int32_t>& axes,
+               const std::string& output_name);
+  void Unsqueeze(const std::string& input_name,
+                 const std::vector<int32_t>& axes,
+                 const std::string& output_name);
+  void AddShape(const std::string& name, const Shape& shape);
+  size_t GetSize(const std::string& name);
+  void Clear();
+
+  inline const Shape& operator[](const std::string& key) {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/shaper.h:125]: >r[] [](start = 29, length = 3)

Semantical mismatch between interface and impl. [] usually does not check the bounds and doesnot throw. However, impl employes at() which does the opposite.

Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/shaper.h:125]: >y) { [](start = 54, length = 4)

Should this be const?

File: onnxruntime/core/providers/rknpu/shaper.h [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+#include <string>
+#include <vector>
+#include <map>
+
+#include <iostream>
+
+namespace onnxruntime {
+namespace rknpu {
+
+class Shaper {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/rknpu/shaper.h:12]: Documentation.
Lots of functions lack constness which may/will lead to const_casts and hard to find bugs.

File: onnxruntime/core/providers/rknpu/node_attr_helper.h (hunk 5 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/node_attr_helper.h]: you means def_val?

File: onnxruntime/core/providers/rknpu/node_attr_helper.h (hunk 6 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/node_attr_helper.h]: of course, const has been added.

File: onnxruntime/core/providers/rknpu/node_attr_helper.h (hunk 7 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/node_attr_helper.h]: added

File: onnxruntime/core/providers/rknpu/node_attr_helper.cc (hunk 8 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/node_attr_helper.cc]: has been modifed.

File: onnxruntime/core/providers/rknpu/node_attr_helper.cc (hunk 9 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/node_attr_helper.cc]: it has been removed.

File: onnxruntime/core/providers/rknpu/onnx_converter.h (hunk 10 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/onnx_converter.h]: onnx_converter is designed to be no-copied, are the following modification correct?

```
private:
  OnnxConverter(const OnnxConverter&);
  OnnxConverter& operator=(const OnnxConverter&);
```

File: onnxruntime/core/providers/rknpu/onnx_converter.h (hunk 11 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/onnx_converter.h]: it has been removed.

File: onnxruntime/core/providers/rknpu/onnx_converter.h (hunk 12 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/onnx_converter.h:72]: std::shared_ptr<rk::nn::Tensor> is used by many interfaces of rknpu_ddk, is there any unreasonable for this?

File: onnxruntime/core/providers/rknpu/onnx_converter.h (hunk 13 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/onnx_converter.h:24]: already added.

File: onnxruntime/core/providers/rknpu/shaper.h (hunk 14 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/shaper.h:172]: is these correct?
```
  inline const Shape& operator[](const std::string& key) {
    return shape_map_[key];
  }
```

File: onnxruntime/core/providers/rknpu/shaper.h (hunk 15 above)
Comment (airockchip) [reply onnxruntime/core/providers/rknpu/shaper.h:15]: do you mean the parameter of functions?

**Comment by jywu-msft** [comment]
build.py has merge conflict. can you please resolve it? thanks.

**Comment by airockchip** [comment]
> build.py has merge conflict. can you please resolve it? thanks.

because of holiday, it was not merged in time, thanks for merged it.
